### PR TITLE
Fixes #11451 - Changed bookmark name to be unique per controller.

### PR DESCRIPTION
**Pull request by alongoldboim**

### CODE REVIEW

File: app/models/bookmark.rb [hunk 1]
```diff
@@ -10,7 +10,7 @@ class Bookmark < ActiveRecord::Base
   attr_accessible :name, :controller, :query, :public
   audited :allow_mass_assignment => true
 
-  validates :name, :uniqueness => true, :unless => Proc.new{|b| Bookmark.my_bookmarks.where(:name => b.name).empty?}
+  validates_uniqueness_of :name, :scope => :controller, :unless => Proc.new{|b| Bookmark.my_bookmarks.where(:name => b.name).empty?}
```
Comment (ohadlevy) [inline app/models/bookmark.rb]: note rubocop failures + missing tests

Comment (alongoldboim) [inline app/models/bookmark.rb]: fixing+adding.

**Comment by alongoldboim** [comment]
@ohadlevy Looks good to me, is this ready for merge?

File: app/models/bookmark.rb [hunk 2]
```diff
@@ -10,7 +10,7 @@ class Bookmark < ActiveRecord::Base
   attr_accessible :name, :controller, :query, :public
   audited :allow_mass_assignment => true
 
-  validates :name, :uniqueness => true, :unless => Proc.new{|b| Bookmark.my_bookmarks.where(:name => b.name).empty?}
+  validates :name,:uniqueness => {:scope => :controller}, :unless => Proc.new{|b| Bookmark.my_bookmarks.where(:name => b.name).empty?}
```
Comment (domcleal) [inline app/models/bookmark.rb]: Why remove the space between hash elements?

File: test/unit/bookmark_test.rb [hunk 3]
```diff
@@ -20,6 +20,15 @@ class BookmarkTest < ActiveSupport::TestCase
     end
   end
 
+  test "my bookmarks should be able to create two bookmarks with same name under different controllers" do
+    enable_login do
```
Comment (domcleal) [inline test/unit/bookmark_test.rb]: I don't think this helper does anything since c6fcfc9.

File: test/unit/bookmark_test.rb [hunk 4]
```diff
@@ -20,6 +20,15 @@ class BookmarkTest < ActiveSupport::TestCase
...
+      assert_difference 'Bookmark.count',2 do
+        Bookmark.create({:name => "private", :query => "bar", :public => false, :controller => "hosts"})
```
Comment (domcleal) [inline test/unit/bookmark_test.rb]: Use `create!` to pick up creation errors, or even just create one (via FG) and then check the second is valid without saving it, to reduce DB load.

File: test/unit/bookmark_test.rb (hunk 3 above)
Comment (alongoldboim) [inline test/unit/bookmark_test.rb]: Your right, I've removed it.
Should i remove it for the rest of the tests on bookmark?

Comment (domcleal) [inline test/unit/bookmark_test.rb]: Sure, delete it.

**Comment by alongoldboim** [comment]
@domcleal fixed.

File: test/unit/bookmark_test.rb [hunk 5]
```diff
@@ -6,17 +6,22 @@ class BookmarkTest < ActiveSupport::TestCase
   end
 
   test "my bookmarks should not contain private bookmarks" do
-    enable_login do
-      assert_equal Bookmark.my_bookmarks.include?(bookmarks(:two)), false
-    end
+    User.current = users(:one)
```
Comment (domcleal) [inline test/unit/bookmark_test.rb]: Use the `as_user` helper which takes a block, then it won't leave User.current set.

File: test/unit/bookmark_test.rb [hunk 6]
```diff
@@ -6,17 +6,22 @@ class BookmarkTest < ActiveSupport::TestCase
...
+    assert_equal Bookmark.my_bookmarks.include?(bookmarks(:two)), false
   end
 
   test "my bookmarks should contain my private bookmarks" do
-    enable_login do
       assert_difference('Bookmark.count') do
```
Comment (domcleal) [inline test/unit/bookmark_test.rb]: Remove 2 spaces of indentation.

File: test/factories/bookmark.rb [hunk 7]
```diff
@@ -0,0 +1,8 @@
+FactoryGirl.define do
+  factory :bookmark do
+    sequence(:name) { |n| "bookmark_#{n}" }
+    query "bar"
+    public false
+  end
+
```
Comment (domcleal) [inline test/factories/bookmark.rb]: Remove the blank line

**Comment by alongoldboim** [comment]
@domcleal ok now?

**Comment by dLobatog** [comment]
Merged as e90e2b56defbf4cc7d7c4bc8ff4cddf74e704a1c, thanks @alongoldboim!

**Comment by dLobatog** [comment]
Thanks @domcleal for reviewing too

**Pull request closed without merging**